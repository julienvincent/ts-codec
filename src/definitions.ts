export type CommonMetadata = {
  description?: string;
  [key: string]: any;
};

export type CodecProps = {
  metadata?: CommonMetadata;
  required?: boolean;
  [key: string]: any;
};

export type Transformer<I, O> = (data: I) => O;

export type Codec<I, O, T = string, P = CodecProps> = {
  _tag: T;
  props: CodecProps & P;

  meta: (metadata: CommonMetadata) => Codec<I, O, T, P>;

  encode: Transformer<I, O>;
  decode: Transformer<O, I>;

  and: <T extends AnyCodec>(codec: T) => Intersection<Codec<I, O>, T>;
  or: <T extends AnyCodec>(codec: T) => Union<Codec<I, O>, T>;

  optional: () => OptionalCodec<Codec<I, O>>;
};

export type PassThroughCodec<T> = Codec<T, T>;

export type AnyCodec = Codec<any, any, any, any>;

export type Cx<C extends AnyCodec> =
  C extends Codec<infer I, infer O, infer T, infer P> ? { I: I; O: O; T: T; P: P } : never;
export type Ix<C extends AnyCodec> = Cx<C>['I'];
export type Ox<C extends AnyCodec> = Cx<C>['O'];

export type Encoded<T extends AnyCodec> = Ox<T>;
export type Decoded<T extends AnyCodec> = Ix<T>;

export type AnyObjectCodecShape = Record<string, AnyCodec>;

type OptionalKeys<T extends AnyObjectCodecShape> = {
  [K in keyof T]: T[K] extends OptionalCodec<any> ? K : never;
}[keyof T];

type RequiredKeys<T extends AnyObjectCodecShape> = Exclude<keyof T, OptionalKeys<T>>;

type WithOptional<T extends AnyObjectCodecShape> = {
  [K in RequiredKeys<T>]: T[K];
} & {
  [K in OptionalKeys<T>]?: T[K];
};

type MappedIx<T extends AnyObjectCodecShape> = { [K in keyof T]: Ix<T[K]> };
type MappedOx<T extends AnyObjectCodecShape> = { [K in keyof T]: Ox<T[K]> };

export enum CodecType {
  String = 'string',
  Number = 'number',
  Boolean = 'boolean',
  Literal = 'literal',
  Enum = 'enum',
  Null = 'null',
  Any = 'any',

  Optional = 'optional',

  Object = 'object',
  Record = 'record',
  Array = 'array',
  Tuple = 'tuple',

  Recursive = 'recursive',

  Union = 'union',
  Intersection = 'intersection'
}

export type Intersection<C1 extends AnyCodec, C2 extends AnyCodec> = Codec<
  Ix<C1> & Ix<C2>,
  Ox<C1> & Ox<C2>,
  CodecType.Intersection,
  {
    codecs: AnyCodec[];
  }
>;

export type Union<C1 extends AnyCodec, C2 extends AnyCodec> = Codec<
  Ix<C1> | Ix<C2>,
  Ox<C1> | Ox<C2>,
  CodecType.Union,
  {
    codecs: AnyCodec[];
  }
>;

export type ObjectCodec<T extends AnyObjectCodecShape> = Codec<
  MappedIx<WithOptional<T>>,
  MappedOx<WithOptional<T>>,
  CodecType.Object,
  {
    shape: T;
  }
>;

export type RecordCodec<T extends AnyCodec> = Codec<
  Record<string, Ix<T>>,
  Record<string, Ox<T>>,
  CodecType.Record,
  {
    type: T;
  }
>;

export type ArrayCodec<T extends AnyCodec> = Codec<
  Ix<T>[],
  Ox<T>[],
  CodecType.Array,
  {
    type: T;
  }
>;

export type AnyTuple = [AnyCodec, ...AnyCodec[]];

export type AssertArray<T> = T extends any[] ? T : never;
export type TupleIx<T extends AnyTuple> = AssertArray<{
  [k in keyof T]: T[k] extends AnyCodec ? Ix<T[k]> : never;
}>;
export type TupleOx<T extends AnyTuple> = AssertArray<{
  [k in keyof T]: T[k] extends AnyCodec ? Ox<T[k]> : never;
}>;

export type TupleCodec<T extends AnyTuple> = Codec<
  TupleIx<T>,
  TupleOx<T>,
  CodecType.Tuple,
  {
    codecs: T;
  }
>;

export type RecursiveCodec<T extends AnyCodec> = Codec<
  Ix<T>,
  Ox<T>,
  CodecType.Recursive,
  {
    id: string;
    resolver: () => T;
  }
>;

export type OptionalCodec<T extends AnyCodec> = Codec<
  Ix<T> | undefined,
  Ox<T> | undefined,
  CodecType.Optional,
  Cx<T>['P']
>;

type IdentityMapping<T extends CodecType> = T extends CodecType.String
  ? string
  : T extends CodecType.Number
    ? number
    : T extends CodecType.Boolean
      ? boolean
      : never;

export type IdentityCodec<T extends CodecType> = Codec<IdentityMapping<T>, IdentityMapping<T>, T>;

export type LiteralCodec<T extends string | number | boolean | null> = Codec<
  T,
  T,
  CodecType.Literal,
  {
    value: T;
  }
>;

export type EnumLike = {
  [key: string]: string;
};
export type EnumCodec<T extends EnumLike> = Codec<
  T[keyof T],
  T[keyof T],
  CodecType.Enum,
  {
    enum: EnumLike;
  }
>;
