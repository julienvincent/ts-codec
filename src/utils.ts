import * as defs from './definitions';

export class TransformError extends Error {
  constructor(readonly errors: string | string[]) {
    super(Array.isArray(errors) ? errors.join(', ') : errors);
    this.name = 'TransformError';
  }

  get errorsArray() {
    return Array.isArray(this.errors) ? this.errors : [this.errors];
  }
}

export function isCodecType<T extends defs.CodecType.String | defs.CodecType.Number | defs.CodecType.Boolean>(
  codec: defs.AnyCodec,
  type: T
): codec is defs.IdentityCodec<T>;
export function isCodecType(codec: defs.AnyCodec, type: defs.CodecType.Object): codec is defs.ObjectCodec<any>;
export function isCodecType(
  codec: defs.AnyCodec,
  type: defs.CodecType.Intersection
): codec is defs.Intersection<defs.AnyCodec, defs.AnyCodec>;
export function isCodecType(
  codec: defs.AnyCodec,
  type: defs.CodecType.Union
): codec is defs.Union<defs.AnyCodec, defs.AnyCodec>;
export function isCodecType(codec: defs.AnyCodec, type: defs.CodecType) {
  return codec._tag === type;
}
