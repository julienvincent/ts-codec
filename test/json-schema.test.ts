import { describe, test, expect } from 'vitest';
import * as t from '../src';

describe('json-schema', () => {
  const date = t.codec<Date, string>(
    'Date',
    (date) => date.toISOString(),
    (date) => new Date(date)
  );

  const A = t.object({
    a: t.string
  });

  const B = t.object({
    b: t.string
  });

  const C = t.object({
    c: t.string
  });

  const D = t.object({
    d: t.string
  });

  test('it should handle a simple identity schema', () => {
    enum SomeEnum {
      A = 'a',
      B = 'b'
    }

    const identity_schema = t.object({
      string: t.string,
      number: t.number,
      boolean: t.boolean,
      any: t.any,
      null: t.Null,
      literal_string: t.literal('literal'),
      literal_num: t.literal(1),
      literal_bool: t.literal(true),
      enum: t.Enum(SomeEnum),
      tuple: t.tuple([t.string, t.number]),
      multi: t.string.or(t.number).or(t.boolean),
      optional: t.string.optional(),

      array: t.array(t.string),
      deeper_array: t.array(t.object({ a: t.string })),

      record: t.record(t.string),

      intersection: t
        .object({
          a: t.string
        })
        .and(
          t
            .object({
              b: t.number
            })
            .or(
              t.object({
                c: t.number
              })
            )
        ),

      with_description: t.string.meta({
        description: 'This is a description'
      })
    });

    expect(t.generateJSONSchema(identity_schema)).toMatchSnapshot();
  });

  test('it should generate schemas from dual codecs', () => {
    const schema = t.object({
      created_at: date
    });

    const DateParser = t.createParser('Date', (codec, options) => {
      switch (options.target) {
        case t.TransformTarget.Encoded: {
          return {
            type: 'string'
          };
        }

        case t.TransformTarget.Decoded: {
          return {
            nodeType: 'date'
          };
        }
      }
    });

    expect(
      t.generateJSONSchema(schema, { target: t.TransformTarget.Encoded, parsers: [DateParser] })
    ).toMatchSnapshot();
    expect(
      t.generateJSONSchema(schema, { target: t.TransformTarget.Decoded, parsers: [DateParser] })
    ).toMatchSnapshot();
  });

  test('it should generate schemas from recursive codecs', () => {
    type Schema = {
      a: string;
      b?: Schema;
    };

    const RecursiveCodec: t.Codec<Schema, Schema> = t.recursive('RecursiveCodec', () => {
      return t.object({
        a: t.string,
        b: RecursiveCodec.optional()
      });
    });

    expect(t.generateJSONSchema(RecursiveCodec)).toMatchSnapshot();
  });

  test('it should merge intersections', () => {
    expect(t.generateJSONSchema(A.and(B).and(C))).toMatchSnapshot();
    expect(t.generateJSONSchema(A.and(A))).toMatchSnapshot();
  });

  test('it should product intersections containing union and object schemas', () => {
    expect(t.generateJSONSchema(B.or(C).and(A.and(D)))).toMatchSnapshot();
  });

  test('it should intersect a set of only unions', () => {
    expect(t.generateJSONSchema(A.or(B).and(C.or(D)))).toMatchSnapshot();
  });
});
