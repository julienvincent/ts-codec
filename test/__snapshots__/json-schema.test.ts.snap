// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`json-schema > it should generate schemas from dual codecs 1`] = `
{
  "additionalProperties": false,
  "definitions": {},
  "properties": {
    "created_at": {
      "type": "string",
    },
  },
  "required": [
    "created_at",
  ],
  "type": "object",
}
`;

exports[`json-schema > it should generate schemas from dual codecs 2`] = `
{
  "additionalProperties": false,
  "definitions": {},
  "properties": {
    "created_at": {
      "nodeType": "date",
    },
  },
  "required": [
    "created_at",
  ],
  "type": "object",
}
`;

exports[`json-schema > it should generate schemas from recursive codecs 1`] = `
{
  "$ref": "#/definitions/RecursiveCodec",
  "definitions": {
    "RecursiveCodec": {
      "additionalProperties": false,
      "properties": {
        "a": {
          "type": "string",
        },
        "b": {
          "$ref": "#/definitions/RecursiveCodec",
        },
      },
      "required": [
        "a",
      ],
      "type": "object",
    },
  },
}
`;

exports[`json-schema > it should handle a simple identity schema 1`] = `
{
  "additionalProperties": false,
  "definitions": {},
  "properties": {
    "any": {},
    "array": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
    "boolean": {
      "type": "boolean",
    },
    "deeper_array": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "a": {
            "type": "string",
          },
        },
        "required": [
          "a",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "enum": {
      "enum": [
        "a",
        "b",
      ],
      "type": "string",
    },
    "intersection": {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "type": "string",
            },
            "b": {
              "type": "number",
            },
          },
          "required": [
            "b",
            "a",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "type": "string",
            },
            "c": {
              "type": "number",
            },
          },
          "required": [
            "c",
            "a",
          ],
          "type": "object",
        },
      ],
    },
    "literal_bool": {
      "const": true,
      "type": "boolean",
    },
    "literal_num": {
      "const": 1,
      "type": "number",
    },
    "literal_string": {
      "const": "literal",
      "type": "string",
    },
    "multi": {
      "anyOf": [
        {
          "type": "string",
        },
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
    "null": {
      "type": "null",
    },
    "number": {
      "type": "number",
    },
    "optional": {
      "type": "string",
    },
    "record": {
      "additionalProperties": {
        "type": "string",
      },
      "properties": {},
      "required": [],
      "type": "object",
    },
    "string": {
      "type": "string",
    },
    "tuple": {
      "items": [
        {
          "type": "string",
        },
        {
          "type": "number",
        },
      ],
      "maxItems": 2,
      "minItems": 2,
      "type": "array",
    },
    "with_description": {
      "description": "This is a description",
      "type": "string",
    },
  },
  "required": [
    "string",
    "number",
    "boolean",
    "any",
    "null",
    "literal_string",
    "literal_num",
    "literal_bool",
    "enum",
    "tuple",
    "multi",
    "array",
    "deeper_array",
    "record",
    "intersection",
    "with_description",
  ],
  "type": "object",
}
`;

exports[`json-schema > it should intersect a set of only unions 1`] = `
{
  "allOf": [
    {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "type": "string",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "b": {
              "type": "string",
            },
          },
          "required": [
            "b",
          ],
          "type": "object",
        },
      ],
    },
    {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "c": {
              "type": "string",
            },
          },
          "required": [
            "c",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "d": {
              "type": "string",
            },
          },
          "required": [
            "d",
          ],
          "type": "object",
        },
      ],
    },
  ],
  "definitions": {},
}
`;

exports[`json-schema > it should merge intersections 1`] = `
{
  "additionalProperties": false,
  "definitions": {},
  "properties": {
    "a": {
      "type": "string",
    },
    "b": {
      "type": "string",
    },
    "c": {
      "type": "string",
    },
  },
  "required": [
    "a",
    "b",
    "c",
  ],
  "type": "object",
}
`;

exports[`json-schema > it should merge intersections 2`] = `
{
  "additionalProperties": false,
  "definitions": {},
  "properties": {
    "a": {
      "type": "string",
    },
  },
  "required": [
    "a",
  ],
  "type": "object",
}
`;

exports[`json-schema > it should product intersections containing union and object schemas 1`] = `
{
  "anyOf": [
    {
      "additionalProperties": false,
      "properties": {
        "a": {
          "type": "string",
        },
        "b": {
          "type": "string",
        },
        "d": {
          "type": "string",
        },
      },
      "required": [
        "b",
        "a",
        "d",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "a": {
          "type": "string",
        },
        "c": {
          "type": "string",
        },
        "d": {
          "type": "string",
        },
      },
      "required": [
        "c",
        "a",
        "d",
      ],
      "type": "object",
    },
  ],
  "definitions": {},
}
`;
